_posts/2022-06-10-slick-intro.md

```diff
@@ -0,0 +1,377 @@
+To generate the DDL scripts, we need to first collect all the slick tables in a collection. 
+
+```scala
+val tables = Seq(movieTable, actorTable, movieActorMappingTable, streamingProviderMappingTable)
+```
+
+Then we can combine them into a slick DDL schema using:
+```scala
+val ddl: profile.DDL = tables.map(_.schema).reduce(_ ++ _)
+```
+Now, we can invoke the method to generate the scripts:
+```scala
+SlickTables.ddl.createIfNotExistsStatements.mkString(";\n")
+```
+
+## 6. Slick-Pg for Postgres
+PostgreSQL has additional powerful datatypes and features. We can use a thirdparty library [slick-pg](https://github.com/tminglei/slick-pg) to use those features in Slick with ease. We have already added the necessary dependencies in the _build.sbt_. 
+
+To use it, we need to write a custom Postgres Profile and use it instead of the slick provided _PostgresProfile_. We can mix-in the traits from slik-pg based on the required features of postgres. Let's add the support for JSON, HStore and Array datatypes.
+
+```scala
+trait CustomPostgresProfile
+    extends ExPostgresProfile with PgJsonSupport with PgPlayJsonSupport 
+    with PgArraySupport with PgHStoreSupport with PgDate2Support {
+  override def pgjson = "jsonb"
+  override protected def computeCapabilities: Set[slick.basic.Capability] =
+    super.computeCapabilities + slick.jdbc.JdbcCapabilities.insertOrUpdate
+
+  override val api = CustomPGAPI
+  object CustomPGAPI
+      extends API
+      with JsonImplicits
+      with HStoreImplicits
+      with ArrayImplicits
+      with DateTimeImplicits {
+    implicit val strListTypeMapper = new SimpleArrayJdbcType[String]("text").to(_.toList)
+    implicit val playJsonArrayTypeMapper =
+      new AdvancedArrayJdbcType[JsValue](
+        pgjson,
+        (s) => utils.SimpleArrayUtils.fromString[JsValue](Json.parse(_))(s).orNull,
+        (v) => utils.SimpleArrayUtils.mkString[JsValue](_.toString())(v)
+      ).to(_.toList)
+  }
+}
+object CustomPostgresProfile extends CustomPostgresProfile
+```
+Now, we can use `CustomPostgresProfile` instead of the `PostgresProfile` to make use of these features.
+
+### 6.1. Querying from an Array Column
+
+Let's see how we can use filter query on an postgres array column. We have created a _MovieLocationsTable_ which has a _movieLocations_ array field. If we want to filter movies which was shot on any of the input locations, we can do that as:
+```scala
+val locations: List[String] = List("location1", "location2")
+val query = SpecialTables.movieLocationsTable.filter(_.locations @& locations.bind)
+```
+The operator `@&` will return true if there is an overlap between the input list and the database column. There are many other such operators on array field, but we are not going to go through all of it here.
```

Comment:
We do need to add a few of the most popular operators and explain what they do in a short list of the form
- "operator ^$#&@* does such and such"

Otherwise, the reader will feel frustrated that they don't understand the structure and how to extend it to other operators.